### ISSUE TITLE
[ATTENTION] Simplistic Attention Mask Application Compromises Transformer Correctness

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#333]
# [ATTENTION] Simplistic Attention Mask Application Compromises Transformer Correctness

## Problem Description

The `BitNetAttention::compute_attention` function in `crates/bitnet-inference/src/layers/attention.rs` implements a simplified attention mask application that lacks proper mask value handling and advanced masking scenarios. The current implementation uses a naive broadcast addition approach that can lead to incorrect attention patterns and numerical instability.

## Environment

- **Affected Component**: BitNet Inference Engine - Multi-Head Attention Layer
- **File**: `/crates/bitnet-inference/src/layers/attention.rs`
- **Function**: `BitNetAttention::compute_attention` (lines 549-599)
- **Feature Flags**: Affects both `cpu` and `gpu` inference modes
- **MSRV**: Rust 1.90.0+

### Root Cause Analysis

The problematic code segment:

```rust
// Apply attention mask if provided
let masked_scores = if let Some(mask) = attention_mask {
    let mask_candle = mask.to_candle()?;
    let _mask_value = -1e9; // Large negative value for masked positions
    scaled_scores.broadcast_add(&mask_candle).context("Failed to apply attention mask")?
} else {
    scaled_scores
};
```

**Critical Issues Identified:**

1. **Unused Mask Value**: The `_mask_value = -1e9` is declared but never used, indicating incomplete implementation
2. **Incorrect Mask Application**: Direct `broadcast_add` assumes the mask contains pre-computed negative values, which is non-standard
3. **Missing Edge Case Handling**: No validation for mask dimensions, data types, or value ranges
4. **Numerical Instability Risk**: No consideration for gradient flow through masked positions
5. **Lack of Mask Type Support**: Cannot handle different mask formats (binary, additive, multiplicative)

### Impact Assessment

**Severity**: High - Affects core transformer attention mechanism correctness

**Affected Functionality:**
- Causal masking for autoregressive generation
- Padding mask handling for variable-length sequences
- Cross-attention masking in encoder-decoder architectures
- Attention pattern visualization and debugging

**Performance Impact:**
- Potential numerical overflow/underflow in attention scores
- Incorrect probability distributions after softmax
- Degraded model quality and generation coherence

## Reproduction Steps

1. Create attention layer with mask:
   ```rust
   let attention_mask = BitNetAttention::create_causal_mask(seq_len, &device)?;
   let output = attention.forward(&hidden_states, Some(&attention_mask), None, None, 0).await?;
   ```

2. Observe that mask values are added directly rather than properly applied
3. Check attention probability distributions - masked positions will have non-zero probabilities

**Expected Result**: Masked positions should have exactly zero probability after softmax
**Actual Result**: Masked positions have small but non-zero probabilities due to incorrect masking

### Primary Implementation

Replace the current masking logic with proper conditional masking:

```rust
// Apply attention mask if provided
let masked_scores = if let Some(mask) = attention_mask {
    let mask_candle = mask.to_candle()?;

    // Validate mask dimensions
    let mask_shape = mask_candle.shape();
    let scores_shape = scaled_scores.shape();
    if !Self::are_mask_dimensions_compatible(mask_shape, scores_shape) {
        return Err(anyhow::anyhow!(
            "Mask shape {:?} incompatible with scores shape {:?}",
            mask_shape.dims(),
            scores_shape.dims()
        ));
    }

    // Apply mask using where_cond for proper masking
    let mask_value = -f32::INFINITY; // Use -inf for complete masking
    let mask_fill = scaled_scores.full_like(mask_value as f64)?;

    // Binary mask: 1 for attend, 0 for mask
    mask_candle.where_cond(&scaled_scores, &mask_fill)?
} else {
    scaled_scores
};
```

### Alternative Approaches

**Option A: Additive Masking (Current Standard)**
```rust
let masked_scores = if let Some(mask) = attention_mask {
    let mask_candle = mask.to_candle()?;
    let large_negative = -1e9f64;
    let additive_mask = mask_candle.affine(large_negative, 0.0)?;
    scaled_scores.add(&additive_mask)?
} else {
    scaled_scores
};
```

**Option B: Multiplicative Masking**
```rust
let masked_scores = if let Some(mask) = attention_mask {
    let mask_candle = mask.to_candle()?;
    // Element-wise multiplication (0 masks, 1 preserves)
    scaled_scores.mul(&mask_candle)?
} else {
    scaled_scores
};
```

### Phase 1: Core Mask Application (Priority: Critical)
- [ ] Implement proper `where_cond` based masking
- [ ] Add mask dimension validation
- [ ] Replace hardcoded -1e9 with configurable mask values
- [ ] Add mask type detection (binary vs additive)

### Phase 2: Advanced Masking Support (Priority: High)
- [ ] Support for 2D causal masks
- [ ] Support for 3D attention masks (batch-specific)
- [ ] Support for 4D masks (head-specific masking)
- [ ] Add mask broadcasting utilities

### Phase 3: Performance Optimization (Priority: Medium)
- [ ] CUDA kernel optimization for masked attention
- [ ] SIMD optimizations for CPU masking
- [ ] Memory-efficient mask storage
- [ ] Attention pattern caching for repeated masks

### Phase 4: Testing & Validation (Priority: High)
- [ ] Unit tests for different mask types
- [ ] Integration tests with real transformer models
- [ ] Cross-validation against reference implementations
- [ ] Performance benchmarks for masked vs unmasked attention

### Unit Tests Required
```rust
#[cfg(test)]
mod attention_mask_tests {
    use super::*;

    #[tokio::test]
    async fn test_causal_mask_zeros_future_positions() {
        // Verify causal mask properly zeros future attention
    }

    #[tokio::test]
    async fn test_padding_mask_ignores_pad_tokens() {
        // Verify padding tokens receive zero attention
    }

    #[tokio::test]
    async fn test_mask_dimension_validation() {
        // Test various mask shape combinations
    }

    #[tokio::test]
    async fn test_numerical_stability() {
        // Verify no overflow/underflow with extreme values
    }
}
```

### Integration Tests
- Cross-validation with C++ BitNet reference implementation
- Comparison with HuggingFace Transformers attention
- End-to-end generation quality assessment

### Performance Benchmarks
- Masked vs unmasked attention throughput
- Memory usage with different mask types
- GPU vs CPU performance comparison

## Related Issues

This issue is part of a broader initiative to eliminate simulation implementations:
- **Related**: CPU Parallel Attention Simplification (#TBD)
- **Related**: GQA Implementation Completeness (#TBD)
- **Depends on**: Quantization Kernel Optimization (#TBD)

### Definition of Done
- [ ] Attention masks are applied correctly using conditional logic
- [ ] All mask types (causal, padding, custom) work properly
- [ ] Masked positions have exactly zero probability after softmax
- [ ] No numerical instability with extreme mask values
- [ ] Comprehensive test coverage (>95% for masking logic)
- [ ] Performance regression <5% for unmasked attention
- [ ] Cross-validation passes with reference implementations

### Validation Requirements
- [ ] All existing attention tests continue to pass
- [ ] New mask-specific tests achieve 100% pass rate
- [ ] Generated text quality maintains or improves
- [ ] Memory usage remains within acceptable bounds
- [ ] GPU/CPU feature flag compatibility maintained

## Risk Assessment

**Technical Risks:**
- Performance regression due to conditional operations
- GPU kernel complexity increase
- Backward compatibility with existing mask formats

**Mitigation Strategies:**
- Staged rollout with feature flags
- Comprehensive benchmarking before merge
- Fallback to original implementation if needed
- Extensive cross-validation testing

## Additional Context

This fix is essential for production-ready transformer inference and impacts:
- Model accuracy and quality
- Numerical stability
- Compatibility with standard transformer architectures
- Performance optimization opportunities

The current implementation represents a significant gap in transformer correctness that must be addressed for reliable neural network inference.
